nengo/spa/spa_ast.py

```diff
@@ -0,0 +1,994 @@
+
+    Attributes
+    ----------
+    root_module : :class:`spa.Module`
+        The root module the encapsulated all of the constructed structures.
+    cortical : :class:`spa.Cortical`
+        Module to manage cortical rules.
+    bg : :class:`spa.BasalGanglia`
+        Module to manage the basal ganglia part of action selection.
+    thalamus : :class:`spa.Thalamus`
+        Module to manage the thalamus part of action selection.
+    sink : :class:`Sink`
+        Node in the AST where some result will be send to.
+    active_net : class:`nengo.Network`
+        Network to add constructed components to.
+    """
+    __slots__ = [
+        'root_module', 'cortical', 'bg', 'thalamus', 'sink', 'active_net']
+
+    def __init__(
+            self, root_module, cortical=None, bg=None, thalamus=None,
+            sink=None, active_net=None):
+        self.root_module = root_module
+        self.cortical = cortical
+        self.bg = bg
+        self.thalamus = thalamus
+        self.sink = sink
+        if active_net is None:
+            active_net = root_module
+        self.active_net = active_net
+
+    def subcontext(self, sink=None, active_net=None):
+        if sink is None:
+            sink = self.sink
+        if active_net is None:
+            active_net = self.active_net
+        return self.__class__(
+            root_module=self.root_module, cortical=self.cortical, bg=self.bg,
+            thalamus=self.thalamus, sink=sink, active_net=active_net)
+
+    @property
+    def sink_module(self):
+        return self.root_module.get_module(self.sink.name, strip_output=True)
+
+    @property
+    def sink_input(self):
+        return self.root_module.get_module_input(self.sink.name)
+
+
+class Type(object):
+    """Describes a type.
+
+    Each part of the AST evaluates to some type.
+    """
+    def __init__(self, name):
+        self.name = name
+
+    def __repr__(self):
+        return '{}({!r})'.format(self.__class__.__name__, self.name)
+
+    def __str__(self):
+        return self.name
+
+    def __eq__(self, other):
+        return self.__class__ is other.__class__ and self.name == other.name
+
+    def __ne__(self, other):
+        return not self == other
+
+
+TAction = Type('TAction')
+TScalar = Type('TScalar')
+TEffect = Type('TEffect')
+TEffects = Type('TEffects')
+
+
+class TVocabulary(Type):
+    """Each vocabulary is treated as its own type.
+
+    All vocabulary types constitute a type class.
+    """
+    def __init__(self, vocab):
+        super(TVocabulary, self).__init__('TVocabulary')
+        self.vocab = vocab
+
+    def __repr__(self):
+        return '{}({!r}, {!r})'.format(
+            self.__class__.__name__, self.name, self.vocab)
+
+    def __str__(self):
+        return '{}<{}>'.format(self.name, self.vocab)
+
+    def __eq__(self, other):
+        return (super(TVocabulary, self).__eq__(other) and
+                self.vocab is other.vocab)
+
+
+class Artifact(object):
```

Comment:
The wiki article isn't very clear on this, but the term has connections to compilers and in particular the byproducts of a build process.